File: services/server/src/server/bets_protocol.py
from typing import Literal
import safe_socket
from lottery import Bet

MSG_TYPE_BET = 0
MSG_TYPE_MULTI_BETS = 1
MSG_TYPE_REQUEST_WINNERS = 2
MSG_TYPE_WINNER = 3
MSG_TYPE_ACK = 4

DELIMITER = "|"
ENCODING = "utf-8"

HEADER_PAYLOAD_LEN_SIZE = 4
HEADER_TYPE_SIZE = 1
HEADER_NO_BETS_SIZE = 4
HEADER_SIZE = HEADER_PAYLOAD_LEN_SIZE + HEADER_TYPE_SIZE
BYTE_ORDER: Literal["big", "little"] = "big"

BET_PAYLOAD_FIELDS_AMOUNT = 6
BET_FIELDS_AMOUNT = 5

class BetsProtocol:
    def __init__(self, socket) -> None:
        self.socket = socket

    def receive_message(self):
        """
        Receives a message from the client.
        """
        payload_len, msg_type = self._receive_header()
        if msg_type is None or payload_len is None:
            return None, None

        if msg_type == MSG_TYPE_MULTI_BETS:
            bets = self._receive_multi_bets(payload_len)
            if bets is None:
                return None, None
            return MSG_TYPE_MULTI_BETS, bets

        if msg_type == MSG_TYPE_BET: # Used for exercise 5
            bet = self._receive_bet(payload_len)
            if bet is None:
                return None, None
            return MSG_TYPE_BET, [bet]

        if msg_type == MSG_TYPE_REQUEST_WINNERS:
            return MSG_TYPE_REQUEST_WINNERS, None

        payload = self._receive_payload(payload_len)
        return msg_type, payload

    def send_winner(self, bet: Bet) -> int | None:
        """
        Sends a winner message to the client.
        """
        msg = self._serialize_bet(bet)
        safe_socket.send_all(self.socket, msg)

    def send_ack(self) -> int | None:
        """
        Sends an acknowledgment message to the client.
        """
        msg = self._build_ack_msg()
        safe_socket.send_all(self.socket, msg)

    def _receive_header(self) -> tuple[int, int] | tuple[None, None]:
        """
        Receives the header of a message from the client.
        """
        header = safe_socket.recv_all(self.socket, HEADER_SIZE)
        if not header or len(header) < HEADER_SIZE:
            return None, None

        payload_len = int.from_bytes(header[:HEADER_PAYLOAD_LEN_SIZE], BYTE_ORDER)
        msg_type = header[HEADER_PAYLOAD_LEN_SIZE]
        return payload_len, msg_type

    def _receive_payload(self, payload_len: int) -> bytes | None:
        """
        Receives the payload of a message from the client.
        """
        if payload_len == 0:
            return b""
        payload = safe_socket.recv_all(self.socket, payload_len)
        if len(payload) < payload_len:
            return None
        return payload

    def _receive_bet(self, payload_len: int) -> Bet | None:
        """
        Receives a single bet from the client.
        """
        payload = self._receive_payload(payload_len)
        if payload is None:
            return None
        return self._deserialize_bet(payload.decode(ENCODING))

    def _receive_multi_bets(self, payload_len: int) -> list[Bet] | None:
        """
        Receives multiple bets from the client.
        """
        num_bets_bytes = safe_socket.recv_all(self.socket, HEADER_NO_BETS_SIZE)
        if not num_bets_bytes or len(num_bets_bytes) < HEADER_NO_BETS_SIZE:
            return None
        number_of_bets = int.from_bytes(num_bets_bytes, BYTE_ORDER)

        payload = self._receive_payload(payload_len)
        if payload is None:
            return None
        bets = self._deserialize_bets(payload.decode(ENCODING), number_of_bets)
        if bets is None or len(bets) != number_of_bets:
            return None
        return bets

    def _build_ack_msg(self) -> bytes:
        """
        Builds an ack message to be sent to the client.
        """
        return (0).to_bytes(HEADER_PAYLOAD_LEN_SIZE, BYTE_ORDER) + bytes([MSG_TYPE_ACK])

    def _deserialize_bet(self, payload: str) -> Bet | None:
        """
        Deserializes a bet from a string.
        """
        try:
            parts = payload.split(DELIMITER)
            if len(parts) != BET_PAYLOAD_FIELDS_AMOUNT:
                return None
            return Bet(
                agency_id=int(parts[0]),
                first_name=parts[1],
                last_name=parts[2],
                document=int(parts[3]),
                birthdate=parts[4],
                number=int(parts[5]),
            )
        except (ValueError, IndexError):
            return None

    def _deserialize_bets(self, payload: str, number_of_bets: int) -> list[Bet] | None:
        """
        Deserializes multiple bets from a string.
        """
        try:
            parts = payload.split(DELIMITER)
            expected_parts = 1 + number_of_bets * BET_FIELDS_AMOUNT
            if len(parts) != expected_parts:
                return None

            agency_id = int(parts[0])
            bets = []

            for i in range(number_of_bets):
                offset = 1 + i * BET_FIELDS_AMOUNT
                bet = Bet(
                    agency_id=agency_id,
                    first_name=parts[offset],
                    last_name=parts[offset + 1],
                    document=int(parts[offset + 2]),
                    birthdate=parts[offset + 3],
                    number=int(parts[offset + 4]),
                )
                bets.append(bet)

            return bets
        except (ValueError, IndexError):
            return None

    def _serialize_bet(self, bet: Bet) -> bytes:
        """
        Serializes a bet into bytes.
        """
        payload = f"{bet.agency_id}{DELIMITER}{bet.first_name}{DELIMITER}{bet.last_name}{DELIMITER}{bet.document}{DELIMITER}{bet.birthdate}{DELIMITER}{bet.number}".encode(ENCODING)
        header = len(payload).to_bytes(HEADER_PAYLOAD_LEN_SIZE, BYTE_ORDER) + bytes([MSG_TYPE_WINNER])
        return header + payload

File: services/client/src/client/bets_protocol.go
package client

import (
	"errors"
	"fmt"
	"io"
	"strconv"
	"strings"
)

const (
	MSG_TYPE_BET             = 0
	MSG_TYPE_MULTI_BETS      = 1
	MSG_TYPE_REQUEST_WINNERS = 2
	MSG_TYPE_WINNER          = 3
	MSG_TYPE_ACK             = 4
	DELIMITER                = "|"

	HEADER_PAYLOAD_LEN_SIZE = 4
	HEADER_TYPE_SIZE        = 1
	HEADER_SIZE             = HEADER_PAYLOAD_LEN_SIZE + HEADER_TYPE_SIZE

	HEADER_NO_BETS_SIZE    = 4                                 // Bytes indicating the number of bets
	HEADER_MULTI_BETS_SIZE = HEADER_SIZE + HEADER_NO_BETS_SIZE // Header size when multiple bets are sent

	BET_PAYLOAD_PARTS_AMOUNT = 6

	MSG_ERROR_COULD_NOT_BUILD_MSG = "Message could not be built"
	MSG_ERROR_DID_NOT_RECEIVE_ACK = "Didn't receive ack message"
)

// BetsProtocol handles the communication protocol for sending and receiving bets over a connection.
type BetsProtocol struct {
	conn Connection
}

// NewBetsProtocol creates a new BetsProtocol instance with the given connection.
func NewBetsProtocol(conn Connection) *BetsProtocol {
	return &BetsProtocol{conn}
}

// SendBet sends a single bet over the connection.
func (betsProtocol *BetsProtocol) SendBet(bet *Bet) error { // Used for exercise 5
	message := betsProtocol.buildBetMessage(bet)
	if message == nil {
		return errors.New(MSG_ERROR_COULD_NOT_BUILD_MSG)
	}

	return betsProtocol.conn.SendAll(message)
}

// SendBets sends multiple bets over the connection.
func (betsProtocol *BetsProtocol) SendBets(bets []*Bet) error {
	message := betsProtocol.buildBetsMessage(bets)
	if message == nil {
		return errors.New(MSG_ERROR_COULD_NOT_BUILD_MSG)
	}

	if err := betsProtocol.conn.SendAll(message); err != nil {
		return err
	}

	return betsProtocol.receiveAck()
}

// ReceiveWinners requests and receives the list of winning bets from the server.
func (betsProtocol *BetsProtocol) ReceiveWinners() ([]*Bet, error) {
	if err := betsProtocol.requestWinners(); err != nil {
		return nil, err
	}

	winners := make([]*Bet, 0)
	for {
		winner, err := betsProtocol.receiveWinner()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err // Discards "partial" winners.
		}

		winners = append(winners, winner)
	}

	return winners, nil
}

// requestWinners sends a request to the server to get the list of winning bets.
func (betsProtocol *BetsProtocol) requestWinners() error {
	message := betsProtocol.buildRequestWinnersMessage()
	return betsProtocol.conn.SendAll(message)
}

// receiveWinner receives a single winning bet from the server.
func (betsProtocol *BetsProtocol) receiveWinner() (*Bet, error) {
	header, err := betsProtocol.receiveHeader()
	if err != nil {
		return nil, err
	}

	payloadLen := extractUint32FromByteArray(header[:HEADER_PAYLOAD_LEN_SIZE])

	payloadBytes, err := betsProtocol.conn.RecvAll(int(payloadLen))
	if err != nil {
		return nil, err
	}

	return betsProtocol.parseBetFromPayload(string(payloadBytes))
}

// receiveAck waits for an ack message from the server after sending bets.
func (betsProtocol *BetsProtocol) receiveAck() error {
	header, err := betsProtocol.receiveHeader()
	if err != nil {
		return errors.New(MSG_ERROR_DID_NOT_RECEIVE_ACK)
	}

	msgType := header[HEADER_PAYLOAD_LEN_SIZE]

	if msgType == MSG_TYPE_ACK {
		return nil
	}

	return errors.New(MSG_ERROR_DID_NOT_RECEIVE_ACK)
}

// receiveHeader receives the header of a message from the server, which includes the payload length and message type.
func (betsProtocol *BetsProtocol) receiveHeader() ([]byte, error) {
	return betsProtocol.conn.RecvAll(HEADER_SIZE)
}

// parseBetFromPayload parses a Bet struct from a payload string received from the server.
func (BetsProtocol) parseBetFromPayload(payload string) (*Bet, error) {
	parts := strings.Split(payload, DELIMITER)
	if len(parts) != BET_PAYLOAD_PARTS_AMOUNT {
		return nil, errors.New(MSG_ERROR_INVALID_LINE)
	}

	agencyId, err := strconv.Atoi(parts[0])
	if err != nil {
		return nil, err
	}

	firstName := parts[1]
	lastName := parts[2]

	document, err := strconv.Atoi(parts[3])
	if err != nil {
		return nil, err
	}

	birthdate := parts[4]

	number, err := strconv.Atoi(parts[5])
	if err != nil {
		return nil, err
	}

	return &Bet{
		agency_id:  agencyId,
		first_name: firstName,
		last_name:  lastName,
		document:   document,
		birthdate:  birthdate,
		number:     number,
	}, nil
}

// formatBetFields formats the fields of a Bet struct into a string to send.
func (BetsProtocol) formatBetFields(bet *Bet) string {
	return fmt.Sprintf("%s%s%s%s%d%s%s%s%d",
		bet.first_name,
		DELIMITER,
		bet.last_name,
		DELIMITER,
		bet.document,
		DELIMITER,
		bet.birthdate,
		DELIMITER,
		bet.number,
	)
}

// buildBetMessage constructs a message to send a single bet over the connection, including the header and payload.
func (betsProtocol BetsProtocol) buildBetMessage(bet *Bet) []byte {
	if bet == nil {
		return nil
	}

	payload := fmt.Sprintf("%d%s%s", bet.agency_id, DELIMITER, betsProtocol.formatBetFields(bet))
	payloadBytes := []byte(payload)
	payloadLen := uint32(len(payloadBytes))

	// [4 bytes: Payload lenght|1 byte: Type|N Bytes: Payload]
	message := make([]byte, HEADER_SIZE+len(payloadBytes))
	insertUint32IntoByteArray(message[0:HEADER_PAYLOAD_LEN_SIZE], payloadLen)
	message[HEADER_PAYLOAD_LEN_SIZE] = MSG_TYPE_BET
	copy(message[HEADER_SIZE:], payloadBytes)

	return message
}

// buildBetsMessage constructs a message to send multiple bets over the connection, including the header and payload.
func (betsProtocol BetsProtocol) buildBetsMessage(bets []*Bet) []byte {
	if len(bets) == 0 {
		return nil
	}

	betStrings := make([]string, len(bets))
	for i, bet := range bets {
		betStrings[i] = betsProtocol.formatBetFields(bet)
	}

	payload := fmt.Sprintf("%d%s%s", bets[0].agency_id, DELIMITER, strings.Join(betStrings, DELIMITER))
	payloadBytes := []byte(payload)
	payloadLen := uint32(len(payloadBytes))
	numberOfBets := uint32(len(bets))

	// [4 bytes: Payload lenght|1 byte: Type|4 bytes: Number of bets|N Bytes: Payload]
	message := make([]byte, HEADER_MULTI_BETS_SIZE+len(payloadBytes))
	insertUint32IntoByteArray(message[0:HEADER_PAYLOAD_LEN_SIZE], payloadLen)
	message[HEADER_PAYLOAD_LEN_SIZE] = MSG_TYPE_MULTI_BETS
	insertUint32IntoByteArray(message[HEADER_SIZE:HEADER_MULTI_BETS_SIZE], numberOfBets)
	copy(message[HEADER_MULTI_BETS_SIZE:], payloadBytes)

	return message
}

// buildRequestWinnersMessage constructs a message to request the list of winning bets from the server.
func (BetsProtocol) buildRequestWinnersMessage() []byte {
	message := make([]byte, HEADER_SIZE)
	insertUint32IntoByteArray(message[0:HEADER_PAYLOAD_LEN_SIZE], 0) // payloadLen = 0 (there's no payload)
	message[HEADER_PAYLOAD_LEN_SIZE] = MSG_TYPE_REQUEST_WINNERS      // Type
	return message
}

// extractUint32FromByteArray extracts a uint32 integer from a byte array.
func extractUint32FromByteArray(array []byte) uint32 {
	return uint32(array[0])<<24 | uint32(array[1])<<16 | uint32(array[2])<<8 | uint32(array[3])
}

// insertUint32IntoByteArray inserts a uint32 integer into a byte array.
func insertUint32IntoByteArray(array []byte, integer uint32) {
	array[0] = byte(integer >> 24)
	array[1] = byte(integer >> 16)
	array[2] = byte(integer >> 8)
	array[3] = byte(integer)
}
